electrode_models/nucleation_single_particle/functions.py

```diff
@@ -0,0 +1,95 @@
+def residual(SV, SVdot, self, sep, counter, params):
+    """
+    Define the residual for the state of the single particle electrode.
+
+    This is an array of differential and algebraic governing equations, one for each state variable in the anode (anode plus a thin layer of electrolyte + separator).
+
+    1. The electric potential in the electrode phase is an algebraic variable.
+        In the anode, phi = 0 is the reference potential for the system.
+        In the cathode, the electric potential must be such that the ionic current is spatially invariant (i.e. it is constant and equal to the external applied current, for galvanostatic simulations).  
+
+        The residual corresponding to these variables (suppose an index 'j') are of the form:
+            resid[j]  = (epression equaling zero)
+
+    2. All other variables are governed by differential equations.
+    
+        We have a means to calculate dSV[j]/dt for a state variable SV[j] (state variable with index j).  
+    
+        The residuals corresponding to these variables will have the form:
+            resid[j] = SVdot[j] - (expression equalling dSV/dt)
+    """
+    import numpy as np
+    import cantera as ct
+    
+    # Initialize the residual:
+    resid = np.zeros((self.nVars,))
+
+    # Save local copies of the solution vectors, pointers for this electrode:
+    SVptr = self.SVptr
+    SV_loc = SV[SVptr['residual']]
+    SVdot_loc = SVdot[SVptr['residual']]
+
+    # Read the electrode and electrolyte electric potential:
+    phi_ed = SV_loc[SVptr['phi_ed']]
+    phi_elyte = phi_ed + SV_loc[SVptr['phi_dl']]
+
+    # Read out electrode bulk composition; set the Cantra object:
+    C_k_ed = SV_loc[SVptr['C_k_ed']] # Molar density (kmol/m3 of phase)
+    X_k_ed = C_k_ed/sum(C_k_ed) # Mole fraction
+    self.bulk_obj.X = X_k_ed
+
+    # Set electric potentials for Cantera objects:
+    self.bulk_obj.electric_potential = phi_ed
+    self.conductor_obj.electric_potential = phi_ed
+    self.elyte_obj.electric_potential = phi_elyte
+    
+    # Faradaic current density is positive when electrons are consumed 
+    # (Li transferred to the anode)
+    sdot_electron = self.surf_obj.get_net_production_rates(self.conductor_obj)
+    i_Far = -ct.faraday*sdot_electron
+    
+    # Double layer current has the same sign as i_Far:
+    i_dl = self.i_ext_flag*params['i_ext']/self.A_surf_ratio - i_Far
```

Comment:
Long term, I'd like to think about whether `A_surf_ratio` is a constant, for this model, or varies.  But this is a fine start, for now.